blackbird_python/blackbird/program.py

```diff
@@ -237,6 +250,37 @@ def __call__(self, **kwargs):
 
                     op['kwargs'][k] = func(**vals)
 
+        # set values for variables and arrays
+        for k, v in prog._var.items(): # pylint: disable=protected-access
+            # it can either be an independent parameter for a variable
+            if isinstance(v, sym.Expr):
+                par = list(v.free_symbols)
+                func = sym.lambdify(par, v)
+
+                try:
+                    vals = {str(p): kwargs[str(p)] for p in par}
+                except KeyError:
+                    raise ValueError("Invalid value for free parameter provided")
+
+                prog._var[k] = func(**vals)
+            # or encapsulated in an array
+            elif isinstance(v, np.ndarray):
+                populated_array = copy.deepcopy(v)
+                for i, row in enumerate(v):
+                    for j, col in enumerate(row):
```

Comment:
Maybe could use [`np.ndindex` for iterating through `v`](https://stackoverflow.com/a/49360371)?

Reply:
Nice, I didn't know about this. Not sure if I think it would be an improvement in this case though. :thinking:

Reply:
Also came across it just now! The improvement for me would be readability (could be subjective), as there are many nested blocks around here and also it would be helpful to see that indeed the loops are used for indexing only

Reply:
:+1: for `np.ndindex`, it's quite neat :)